lib/cocoapods/target/pod_target.rb

```diff
@@ -127,18 +128,30 @@ def label
       end
     end
 
-    # @return [String] the Swift version for the target. If the pod author has provided a Swift version
-    #                  then that is the one returned, otherwise the Swift version is determined by the user
-    #                  targets that include this pod target.
+    # @return [String] the Swift version for the target. If the pod author has provided a set of Swift versions
+    #         supported by their pod then the max Swift version across all of target definitions is chosen, unless
+    #         a target definition specifies explicit requirements for supported Swift versions. Otherwise the Swift
+    #         version is derived by the target definitions that integrate this pod as long as they are the same.
     #
     def swift_version
-      spec_swift_version || target_definitions.map(&:swift_version).compact.uniq.first
+      @swift_version ||= begin
+        if spec_swift_versions.empty?
+          target_definitions.map(&:swift_version).compact.uniq.first
+        else
+          spec_swift_versions.sort.reverse_each.find do |swift_version|
+            target_definitions.all? do |td|
+              td.supports_swift_version?(swift_version)
+            end
+          end.to_s
```

Comment:
out of scope of this PR but eventually this should probably transition to some sort of `SwiftVersion` (or `Version`) type, especially since `swift_versions` returns `Array<Version>`

Reply:
hm I can do in a follow up for `Version`